onnx/version_converter/adapters/q_dq_21_20.h

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) ONNX Project Contributors
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Adapter for Cast in default domain from version 9 to 8
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "onnx/version_converter/adapters/type_restriction.h"
+
+namespace ONNX_NAMESPACE {
+namespace version_conversion {
+
+static const std::vector<TensorProto_DataType> q_dq_20_unallowed_types = {
+    TensorProto_DataType_UINT16,
+    TensorProto_DataType_INT16};
+
+class QuantizeLinear_21_20 final : public TypeRestriction {
+ public:
+  explicit QuantizeLinear_21_20()
+      : TypeRestriction("QuantizeLinear", OpSetID(21), OpSetID(20), q_dq_20_unallowed_types) {}
+
+  void adapt_quantize_linear_21_20(std::shared_ptr<Graph>, Node* node) const {
+    if (node->inputs()[1]->sizes().size() > 1) {
```

Comment:
I think this could be problematic. In general, there is no guarantee that we know the shape/rank of an input tensor. But this check requires it. I wonder if we should rely on an attribute to determine if we are using blocked quantization. This might be another reason to add an attribute (for blocked quantization), being discussed elsewhere.

Reply:
I was operating under the assumption that scale and zero-point have a constant shape, since the existing implementation for per-axis quantization relies on it, see [source code](https://github.com/onnx/onnx/blob/c5d91f68526ac80563ebd0cb1788fc4a5a383723/onnx/reference/ops/op_quantize_linear.py#L75). However, when examining the documentation, I don't see this explicitly mentioned. 

If scale/zero-point have a constant shape, and since the block size and axis are also known parameters, then the input shape is well defined (therefore cannot be dynamic):
```input_shape = scale_shape[:,axis] + scale_shape[axis]*block_size + scale_shape[axis+1,:]```

In any case, I've already agreed to add the `block_size` attribute, but this discussion is still relevant for testing and documentation.